rfcs/9999-new-merkle-storage.md

```diff
@@ -0,0 +1,140 @@
+exploit redundancy if you don't explicitly use masks. For example we expect the
+epoch and snarked ledgers to be very similar, there is probably a power law
+distribution on how active accounts are, the vast majority will not have changed
+at all since the epoch snapshot was taken. But in the current system it'd be
+very difficult to exploit that.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+The system is composed of two layers. At the bottom is the Merkle storage system
+itself. It stores and tracks immutable hashed data, both on disk and in RAM,
+using reference counting to free storage appropriately. There is only one copy
+of any item with a given hash, so sharing is automatic and implicit. The primary
+object in storage is a `Ref` (name pending). A `Ref` represents some hashed
+data which may reference other data by hash, if you have a `Ref`, then you have
+the data with the associated hash. The `Ref` type has two parameters, A) a
+phantom that is either `full` or `partial` indicating whether we are guaranteed
+to have all the referenced data as well, and B) what the type of the data is.
+There is a function that takes a `Ref` and ensures that it and all its
+references are stored on disk. We call this "pinning" the `Ref`.
+
+In the table below, the rows correspond to whether the data is pinned and the
+columns correspond to the value of the "fullness" type parameter. The inner
+elements are what those combinations are in the current system (for ledgers).
+
+|        | Full                             | Partial         |
+|--------|----------------------------------|-----------------|
+| Disk   | `Merkle_ledger.Database`         | N/A             |
+| Memory | `Ledger.Mask.Attached` (to null) | `Sparse_ledger` |
+
+Most (maybe all?) of the in-memory ledgers we use in the daemon are actually
+masks on top of a database ledger, which doesn't fit into the new model
+described in this RFC.
+
+The system for synchronizing Merkle structures over the network exists at this
+level, replacing `syncable_ledger`.
+
+On top of the underlying storage layer are the actual Merkle structures. In
+order to store an Merkle object, the programmer writes a module that describes
+the object, how to (de)serialize it, and how to find its references. They pass
+that object into a functor and get the `Ref` machinery, specialized to the type
+in question, back. When implementing this, we'd write a module that describes
+the complete binary tree that is our ledger - the result of the functor would be
+used to build something close to `Base_ledger_intf.S`. We could also use this
+for a Merkleized parallel scan state, or to replace
+`receipt_chain_database`.
+
+### Storage management
+
+There are two tables mapping hashes to objects: an in-memory hashtable and
+RocksDB on disk. The in-memory table contains *weak* references, and entries are
+removed from it by a finalizer when the associated data is garbage collected. We
```

Comment:
Given that we need to keep track of objects manually for the ones stored on disk, we should just use the manual reference counting for these too, no?

Can we rely on the GC to actually collect when we want it to? We can still use the finalizers the same way we're using them for the masks, to ensure we don't leak things.

Reply:
I think it's easier to not have reference counting logic for the stuff that the GC will track anyway.

The GC isn't guaranteed to collect stuff when we want it to, but it's good at freeing up memory and that's the reason we want this in the first place.